worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala

```diff
@@ -1246,6 +1204,62 @@ class PushDataHandler(val workerSource: WorkerSource) extends BaseMessageHandler
     }
   }
 
+  private def writeLocalData(
+      fileWriters: Seq[FileWriter],
+      body: ByteBuf,
+      shuffleKey: String,
+      isPrimary: Boolean,
+      batchOffsets: Option[Array[Int]],
+      writePromise: Promise[Unit]): Unit = {
+    def writeData(fileWriter: FileWriter, body: ByteBuf, shuffleKey: String): Unit = {
+      try {
+        fileWriter.write(body)
+      } catch {
+        case e: Exception =>
+          if (e.isInstanceOf[AlreadyClosedException]) {
+            val (mapId, attemptId) = getMapAttempt(body)
+            val endedAttempt =
+              if (shuffleMapperAttempts.containsKey(shuffleKey)) {
+                shuffleMapperAttempts.get(shuffleKey).get(mapId)
+              } else -1
+            // TODO just info log for ended attempt
+            logWarning(s"Append data failed for task(shuffle $shuffleKey, map $mapId, attempt" +
+              s" $attemptId), caused by AlreadyClosedException, endedAttempt $endedAttempt, error message: ${e.getMessage}")
+          } else {
+            logError("Exception encountered when write.", e)
+          }
+          if (!writePromise.isCompleted) {
```

Comment:
I think we need to move this check inside the loop over batchOffsets to avoid calling `writeData` for successive partitions if `writePromise.isCompleted`

Reply:
makes sense